pkgs/development/haskell-modules/generic-builder.nix

```diff
@@ -383,11 +383,20 @@ stdenv.mkDerivation ({
       mv tmp "$d"
     done
 
-    for d in $(grep '^dynamic-library-dirs:' "$packageConfDir"/* | cut -d' ' -f2- | tr ' ' '\n' | sort -u); do
-      for lib in "$d/"*.{dylib,so}; do
-        ln -s "$lib" "$dynamicLinksDir"
+    listDynamicLinks () {
+      for d in $(grep '^dynamic-library-dirs:' "$packageConfDir"/* | cut -d' ' -f2- | tr ' ' '\n' | sort -u); do
+        for lib in "$d/"*.{dylib,so} ; do
+          if [ -f "$lib" ] ; then
+            echo "$lib"
+          fi
+        done
```

Comment:
I was thinking of something like:
```shell
  for lib in "$d/"*.{dylib,so} ; do
    local linkPath="$dynamicLinksDir/$(basename "$lib")"
    if [ -e "$linkPath" ]; then
      # link already exists! but maybe that's OK
      if [ ! -l "$linkPath" ] || [ "$(readlink "$linkPath")" != "$lib" ]; then
        echo >&2 "error: failed to create symbolic link $linkPath: file exists"
        exit 1
      fi
    else
      ln -s "$lib" "$linkPath"
    fi
  done
```

I.e., if the place where we want to create the link already exists, then check that it is already a link and already points exactly where we'd like to point it.